File: src/services/downloadPlan.ts
import type { AndroidGameUpdateInfo } from "./gameUpdate";

export type NativeDownloadChunk = {
  index: number;
  count: number;
  fileName: string;
  objectKey: string;
  sha256: string;
  sizeBytes: number;
  downloadUrl?: string;
};

export type AndroidDownloadSource = "official" | "github";

export type AndroidDownloadPlan = {
  productKey: "crossingvoid-android-game";
  runtime: "Android";
  source: AndroidDownloadSource;
  version: string;
  archiveFileName: string;
  archiveSha256: string;
  totalBytes: number;
  chunks: NativeDownloadChunk[];
};

export type LauncherDownloadPhase =
  | "updateReady"
  | "downloading"
  | "paused"
  | "verifying"
  | "readyInstall"
  | "error";

function requireSha256(value: string | undefined, label: string) {
  const normalized = value?.trim().toLowerCase() ?? "";
  if (!/^[a-f0-9]{64}$/.test(normalized)) throw new Error(`${label}缺少有效 SHA256`);
  return normalized;
}

function githubChunkUrl(version: string, fileName: string) {
  const tag = `Android-${version}`;
  return `https://github.com/kirito0000001/CrossingVoid/releases/download/${encodeURIComponent(tag)}/${encodeURIComponent(fileName)}`;
}

export function buildAndroidDownloadPlan(
  update: AndroidGameUpdateInfo,
  source: AndroidDownloadSource = "official",
): AndroidDownloadPlan {
  const chunks = [...update.asset.chunks].sort((left, right) => left.index - right.index);
  if (chunks.length === 0) throw new Error("Android 更新清单没有下载分片");

  const normalized = chunks.map((chunk, position) => {
    const expectedIndex = position + 1;
    if (chunk.index !== expectedIndex) throw new Error(`Android 分片序号不连续：应为 ${expectedIndex}，实际为 ${chunk.index}`);
    if (chunk.count !== chunks.length) throw new Error(`Android 分片数量不一致：清单声明 ${chunk.count}，实际为 ${chunks.length}`);
    if (!chunk.fileName.trim() || !chunk.objectKey.trim()) throw new Error(`Android 第 ${expectedIndex} 个分片缺少文件信息`);
    if (!Number.isSafeInteger(chunk.sizeBytes) || (chunk.sizeBytes ?? 0) <= 0) throw new Error(`Android 第 ${expectedIndex} 个分片大小无效`);

    return {
      index: expectedIndex,
      count: chunks.length,
      fileName: chunk.fileName,
      objectKey: chunk.objectKey,
      sha256: requireSha256(chunk.sha256, `Android 第 ${expectedIndex} 个分片`),
      sizeBytes: chunk.sizeBytes!,
      ...(source === "github" ? { downloadUrl: githubChunkUrl(update.version, chunk.fileName) } : {}),
    };
  });

  const chunkBytes = normalized.reduce((total, chunk) => total + chunk.sizeBytes, 0);
  if (chunkBytes !== update.asset.sizeBytes) {
    throw new Error(`Android 分片总大小 ${chunkBytes} 与完整包 ${update.asset.sizeBytes} 不一致`);
  }

  return {
    productKey: "crossingvoid-android-game",
    runtime: "Android",
    source,
    version: update.version,
    archiveFileName: update.asset.fileName,
    archiveSha256: requireSha256(update.asset.sha256, "Android 完整包"),
    totalBytes: update.asset.sizeBytes,
    chunks: normalized,
  };
}

export function launcherPhaseFromNativeState(status: string): LauncherDownloadPhase {
  if (status === "downloading") return "downloading";
  if (status === "paused") return "paused";
  if (["verifying", "merging", "extracting"].includes(status)) return "verifying";
  if (status === "ready") return "readyInstall";
  if (status === "error") return "error";
  return "updateReady";
}

File: android/app/src/main/java/com/lingjing/launcher/android/DownloadPlan.java
package com.lingjing.launcher.android;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.Locale;

/**
 * 游戏下载清单。
 *
 * 构造时校验产品键、运行平台、下载来源、分片连续性与总大小对账，
 * 以及整包 SHA-256 的格式。清单来自网络，任何一项不成立都直接拒绝。
 */
final class DownloadPlan {
    final String productKey;
    final String runtime;
    final String source;
    final String version;
    final String archiveFileName;
    final String archiveSha256;
    final long totalBytes;
    final List<DownloadChunk> chunks;

    private DownloadPlan(JSONObject source) throws JSONException {
        productKey = source.getString("productKey");
        runtime = source.getString("runtime");
        this.source = source.getString("source");
        version = source.getString("version");
        archiveFileName = source.getString("archiveFileName");
        archiveSha256 = source.getString("archiveSha256").toLowerCase(Locale.ROOT);
        totalBytes = source.getLong("totalBytes");
        chunks = new ArrayList<>();
        JSONArray items = source.getJSONArray("chunks");
        for (int index = 0; index < items.length(); index++) {
            chunks.add(new DownloadChunk(items.getJSONObject(index)));
        }
        chunks.sort(Comparator.comparingInt(chunk -> chunk.index));
        long chunkBytes = 0L;
        for (int position = 0; position < chunks.size(); position++) {
            DownloadChunk chunk = chunks.get(position);
            if (chunk.index != position + 1 || chunk.count != chunks.size()) throw new JSONException("Android 游戏分片序号不连续");
            if (this.source.equals("github") && chunk.downloadUrl.isBlank()) throw new JSONException("Github 游戏分片缺少下载地址");
            chunkBytes = Math.addExact(chunkBytes, chunk.sizeBytes);
        }
        if (!productKey.equals("crossingvoid-android-game") || !runtime.equals("Android")
            || (!this.source.equals("official") && !this.source.equals("github")) || chunks.isEmpty()
            || totalBytes <= 0 || chunkBytes != totalBytes || !archiveSha256.matches("^[a-f0-9]{64}$")) {
            throw new JSONException("下载清单不完整");
        }
    }

    static DownloadPlan parse(String json) throws JSONException {
        return new DownloadPlan(new JSONObject(json));
    }

    boolean matchesState(JSONObject state) {
        return version.equals(state.optString("version")) && archiveSha256.equalsIgnoreCase(state.optString("archiveSha256"));
    }
}
